### ISSUE TITLE
feat: Differentiable MM backend via JAX for analytical gradients

### ISSUE THREAD

**Issue by ericchansen** [ericchansen/q2mm#68]
## Summary

Build a differentiable MM energy backend using JAX so that the objective function gradient with respect to all force field parameters can be computed analytically via automatic differentiation, rather than by finite differences.

## Motivation

Q2MM currently uses **central finite differences** to compute the gradient of the objective function with respect to force field parameters. For N parameters, this requires 2N MM energy evaluations per gradient step:

| Approach | Cost per gradient step (100 params) |
|----------|-------------------------------------|
| Central finite differences (current) | **200 MM evaluations** |
| Forward finite differences | **100 MM evaluations** |
| Reverse-mode autodiff (JAX) | **~1-2 forward passes** |

This is the single biggest computational bottleneck in Q2MM. With analytical gradients, we also get:

- **Exact Hessians** (d^2 chi^2 / d params^2) for free, enabling true Newton methods
- **Full parameter coupling information** -- which parameters interact
- **No eps tuning** -- the current scipy optimizer requires `eps=1e-3` because the default finite-difference step (~1e-8) produces sub-precision energy changes. Analytical gradients eliminate this entirely.

## Background

Per-Ola Norrby and collaborators have long advocated for using exact/functional gradients from the Hessian to accelerate force field optimization. Several groups have demonstrated this approach:

- **DMFF** (Deep Modelling Force Field) -- JAX-based differentiable FF fitting
- **JAX-MD** -- Differentiable molecular dynamics in JAX
- **JAX-ReaxFF** -- Differentiable reactive force fields
- **TorchANI** -- PyTorch-based neural network potentials with autodiff

The common insight: if the MM energy function `E(params, coords)` is written in a differentiable framework, the chain rule gives exact gradients of ANY objective function through the energy computation.

## Current Architecture

The existing OpenMM backend uses:
- `CustomBondForce` with MM3 cubic expansion: `E = k * (dr^2 - 2.55*dr^3 + 7/12*2.55^2*dr^4)`
- `CustomAngleForce` with MM3 quartic expansion
- `CustomNonbondedForce` with buffered 14-7 LJ
- Parameter updates via `setBondParameters()` + `updateParametersInContext()`

OpenMM itself is C++/CUDA and not differentiable. The JAX backend would reimplement these same functional forms in pure JAX.

### Phase 1: JAX Energy Functions

```python
# q2mm/backends/mm/jax_engine.py
import jax
import jax.numpy as jnp

def mm3_bond_energy(params, coords, bond_indices):
    """MM3 cubic bond stretch -- differentiable."""
    k, r0 = params
    dr = distances(coords, bond_indices) - r0
    return jnp.sum(k * (dr**2 - 2.55*dr**3 + 7/12*2.55**2*dr**4))

def mm3_angle_energy(params, coords, angle_indices):
    """MM3 quartic angle bend -- differentiable."""
    ...

def total_energy(all_params, coords, topology):
    """Total MM energy as a pure function of params."""
    return bond_energy + angle_energy + vdw_energy
```

### Phase 2: Differentiable Objective Function

```python
def objective(params, ref_data, molecules, topology):
    """Objective function with analytical gradient."""
    energies = jax.vmap(total_energy, in_axes=(None, 0, None))(params, all_coords, topology)
    residuals = compute_residuals(energies, ref_data)
    return jnp.sum(residuals**2)

grad_fn = jax.grad(objective)        # Exact gradient
hess_fn = jax.hessian(objective)     # Exact Hessian

# One call = gradient w.r.t. ALL parameters
gradient = grad_fn(current_params, ref_data, molecules, topology)
```

### Phase 3: Integration with Scipy

```python
from scipy.optimize import minimize

result = minimize(
    objective,
    x0=initial_params,
    jac=grad_fn,       # Analytical Jacobian
    hess=hess_fn,      # Analytical Hessian (for Newton methods)
    method='trust-ncg' # Can now use Hessian-based methods
)
```

## Performance Projections

Based on published benchmarks from DMFF and JAX-MD:

| Metric | Finite Diff (current) | JAX Autodiff |
|--------|----------------------|--------------|
| Gradient cost | 2N evals | ~1 eval |
| Hessian cost | O(N^2) evals | ~N evals |
| GPU acceleration | No (OpenMM C++) | Yes (XLA) |
| Precision | eps-dependent | Machine precision |
| Memory | O(1) | O(N) for backprop |

For a typical Q2MM problem with ~50 parameters and ~200 reference data points, this could reduce wall-clock time from hours to minutes.

## Acceptance Criteria

- [ ] JAX implementations of MM3 bond, angle, and vdW energy terms
- [ ] Verified numerical agreement with OpenMM backend (sub-millikcal)
- [ ] `jax.grad()` produces correct gradients (verified against finite differences)
- [ ] `JaxEngine` class implementing the MM backend interface
- [ ] Scipy optimizer integration with analytical Jacobian
- [ ] Benchmark: JAX vs OpenMM finite-diff on SN2 test case
- [ ] JAX listed as optional dependency (`pip install q2mm[jax]`)
- [ ] Documentation showing the speedup and usage

## Dependencies

- `jax` and `jaxlib` (MIT license, pip-installable)
- Optional: `jax[cuda]` for GPU acceleration

## References

- DMFF: https://github.com/deepmodeling/DMFF
- JAX-MD: https://github.com/jax-md/jax-md
- JAX: https://github.com/google/jax
- Norrby & Liljefors, J. Comput. Chem. 1998, 19, 1146-1166 (original Q2MM method)
